spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/SkipperAppRegistryController.java

```diff
@@ -230,9 +250,70 @@ public void makeDefault(@PathVariable("type") ApplicationType type, @PathVariabl
 	@ResponseStatus(HttpStatus.OK)
 	public void unregister(@PathVariable("type") ApplicationType type, @PathVariable("name") String name,
 			@PathVariable("version") String version) {
+
+		if (type != ApplicationType.task) {
+			String streamWithApp = findStreamContainingAppOf(type, name, version);
+			if (streamWithApp != null) {
+				throw new UnregisterAppException(String.format("The app [%s:%s:%s] you're trying to unregister is " +
+						"currently used in stream '%s'.", name, type, version, streamWithApp));
+			}
+		}
+
 		appRegistryService.delete(name, type, version);
 	}
 
+	/**
+	 * Given the application type, name, and version, determine if it is being used in a deployed stream definition.
+	 *
+	 * @param appType the application type
+	 * @param appName the application name
+	 * @param appVersion application version
+	 * @return the name of the deployed stream where the app is being used.  If the app is not deployed in a stream,
+	 * return {@code null}.
+	 */
+	private String findStreamContainingAppOf(ApplicationType appType, String appName, String appVersion) {
+		Iterable<StreamDefinition> streamDefinitions = streamDefinitionRepository.findAll();
+		for (StreamDefinition streamDefinition : streamDefinitions) {
+			StreamDeployment streamDeployment = this.streamService.info(streamDefinition.getName());
+			for (StreamAppDefinition streamAppDefinition : streamDefinition.getAppDefinitions()) {
+				final String streamAppName = streamAppDefinition.getRegisteredAppName();
+				final ApplicationType streamAppType;
+				try {
+					streamAppType = DataFlowServerUtil.determineApplicationType(streamAppDefinition);
+				}
+				catch (CannotDetermineApplicationTypeException e) {
+					logger.warn("Can not determine ApplicationType for " + streamAppDefinition);
+					continue;
+				}
+				if (appType != streamAppType) {
+					continue;
+				}
+				Map<String, Map<String, String>> streamDeploymentPropertiesMap;
+				String streamDeploymentPropertiesString = streamDeployment.getDeploymentProperties();
```

Comment:
If the stream is not yet deployed but the definition exists, then the app would still get unregistered. Because, the value for `streamDeploymentPropertiesString` would be `null` as the stream deployment properties is calculated from the Skipper manifest (from the deployed stream).

Reply:
As the commit message says, this PR addresses the prevention of app unregistration only for the `deployed` stream. So, in that perspective, this flow requires a `null` check on `streamDeploymentPropertiesString` to avoid exception processing it further.

Reply:
Yeah, noticed this as well. If starting skipper and dataflow from scratch, created a stream without deploying it, you get:
```
dataflow:>app unregister --name log --type sink 
Command failed org.springframework.cloud.dataflow.rest.client.DataFlowClientException: Can not deserialize Stream Deployment Properties JSON ''
```

Reply:
Once we have a `null` check, I guess the above error would go away as well.

Reply:
I'm adding a new test and:
```
if (!StringUtils.hasText(streamDeploymentPropertiesString)) {
	continue;
}
```

Reply:
I guess #2013 can still be kept open until we confirm on supporting the prevention of unregistration of the app for the stream that is **not** deployed yet.